src/coreclr/jit/gentree.h

```diff
@@ -4830,13 +4835,47 @@ struct GenTreeCall final : public GenTree
 
     GenTreeCall(var_types type) : GenTree(GT_CALL, type)
     {
-        fgArgInfo = nullptr;
+        fgArgInfo   = nullptr;
+        gtRetBufArg = nullptr;
     }
 #if DEBUGGABLE_GENTREE
     GenTreeCall() : GenTree()
     {
     }
 #endif
+
+    GenTree* GetLclRetBufArgNode() const
+    {
+        if (gtRetBufArg == nullptr)
+        {
```

Comment:
Should we assert on both sides of this? (that the flag and the class member agree)
```suggestion
        {
           assert(!HasRetBufArg());
```

Reply:
Turns out that when we get `RetBuf` as argument in a register (as I understand), and if that `RetBuf` is further used in a call in current method, that argument is propagated as-is without having to put it on current frame:

```
               [000000] S-C-G-------              *  CALL      void   hackishModuleName.hackishMethodName
               [000001] ------------ arg0         \--*  LCL_VAR   byref  V01 RetBuf  
```

In such cases, we are not passing the address and hence we don't perform the optimization of this PR. For such scenarios, `gtRetBufArg == nullptr` while `HasRetBufArg()` is true. We could extract the first argument and then validate for such cases if its `varDsc->lvIsRegArg == true`, but not sure if it is worth it.

Reply:
Do you have a simple example where this happens?

Reply:
In the current design `gtRetBufArg` is only set if the buffer's address is local.

(Simple case where it is not: `*globalPtr = CallThatReturnsStruct();`).

This is mainly for throughput reasons -- we'd like to be able to early out of `GetLclRetBufArgNode`.

Reply:
https://github.com/dotnet/performance/blob/d7dac8a7ca12a28d099192f8a901cf8e30361384/src/benchmarks/micro/libraries/System.Numerics.Vectors/Perf_Matrix3x2.cs#L51 is one such example.